File: app/src/main/java/com/wilinz/yuetingmusic/ui/commen/MusicAdapter.java
package com.wilinz.yuetingmusic.ui.commen;

import android.annotation.SuppressLint;
import android.view.LayoutInflater;
import android.view.ViewGroup;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;

import com.bumptech.glide.Glide;
import com.wilinz.yuetingmusic.R;
import com.wilinz.yuetingmusic.data.model.Song;
import com.wilinz.yuetingmusic.databinding.ItemMusicBinding;

import java.util.List;

public class MusicAdapter extends RecyclerView.Adapter<MusicAdapter.MusicViewHolder> {

    private List<Song> songs;
    private OnItemClickListener listener;

    public interface OnItemClickListener {
        void onItemClick(List<Song> songs, int index, Song song);
    }

    public MusicAdapter(@NonNull List<Song> musics) {
        this.songs = musics;
    }

    public void setSongs(List<Song> songs) {
        this.songs = songs;
        notifyDataSetChanged();
    }

    public void setOnItemClickListener(OnItemClickListener l) {
        listener = l;
    }

    class MusicViewHolder extends RecyclerView.ViewHolder {

        ItemMusicBinding binding;

        public MusicViewHolder(@NonNull ItemMusicBinding binding) {
            super(binding.getRoot());
            this.binding = binding;
        }
    }

    @NonNull
    @Override
    public MusicViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        ItemMusicBinding binding = ItemMusicBinding.inflate(LayoutInflater.from(parent.getContext()), parent, false);
        MusicViewHolder holder = new MusicViewHolder(binding);
        binding.getRoot().setOnClickListener(v -> {
            if (listener != null) {
                int index = holder.getAbsoluteAdapterPosition();
                listener.onItemClick(songs, index, songs.get(index));
            }
        });
        return holder;
    }

    @SuppressLint("SetTextI18n")
    @Override
    public void onBindViewHolder(@NonNull MusicViewHolder holder, int position) {
        ItemMusicBinding binding = holder.binding;
        Song song = songs.get(position);
        binding.name.setText(song.title);
        binding.serialNumber.setText((position + 1) + "");
        binding.secondName.setText(song.artist + "-" + song.album);
        Glide.with(binding.songAvatar)
                .load(song.coverImgUrl)
                .error(R.drawable.icon)
                .into(binding.songAvatar);
    }

    @Override
    public int getItemCount() {
        return songs.size();
    }


}

File: app/src/main/java/com/wilinz/yuetingmusic/ui/commen/MusicAdapter.kt
package com.wilinz.yuetingmusic.ui.commen

import android.annotation.SuppressLint
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.recyclerview.widget.RecyclerView
import com.bumptech.glide.Glide
import com.wilinz.yuetingmusic.R
import com.wilinz.yuetingmusic.data.model.Song
import com.wilinz.yuetingmusic.databinding.ItemMusicBinding
import com.wilinz.yuetingmusic.ui.commen.MusicAdapter.MusicViewHolder
import com.wilinz.yuetingmusic.util.ScreenUtil

class MusicAdapter(private var songs: List<Song>) : RecyclerView.Adapter<MusicViewHolder>() {
    private var listener: ((songs: List<Song>, index: Int, song: Song) -> Unit)? = null


    interface OnItemClickListener {
        fun onItemClick(songs: List<Song>?, index: Int, song: Song?)
    }

    fun setSongs(songs: List<Song>) {
        this.songs = songs
        notifyDataSetChanged()
    }

    fun setOnItemClickListener(l: (songs: List<Song>, index: Int, song: Song) -> Unit) {
        listener = l
    }

    inner class MusicViewHolder(var binding: ItemMusicBinding) : RecyclerView.ViewHolder(
        binding.root
    )

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): MusicViewHolder {
        val binding = ItemMusicBinding.inflate(LayoutInflater.from(parent.context), parent, false)
        val holder = MusicViewHolder(binding)
        binding.root.setOnClickListener { v: View? ->
            listener?.invoke(
                songs,
                holder.absoluteAdapterPosition,
                songs[holder.absoluteAdapterPosition]
            )
        }
        return holder
    }

    @SuppressLint("SetTextI18n")
    override fun onBindViewHolder(holder: MusicViewHolder, position: Int) {
        val binding = holder.binding
        val song = songs[position]
        binding.name.text = song.title
        binding.serialNumber.text = (position + 1).toString() + ""
        binding.secondName.text = song.artist + "-" + song.album
        val size = ScreenUtil.dpToPx(binding.songAvatar.context, 48)
        Glide.with(binding.songAvatar)
            .load(song.coverImgUrl + "?param=${size}y${size}")
            .error(R.drawable.icon)
            .into(binding.songAvatar)
    }

    override fun getItemCount(): Int {
        return songs.size
    }
}